malariagen_data/anopheles.py

```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            sample_query=sample_query,
+            sample_sets=sample_sets,
+            site_mask=site_mask,
+            random_seed=random_seed,
         )
 
-        title_lines = []
-        if sample_sets is not None:
-            title_lines.append(f"sample sets: {sample_sets}")
-        if sample_query is not None:
-            title_lines.append(f"sample query: {sample_query}")
-        title_lines.append(f"genomic region: {region} ({ht.shape[0]} SNPs)")
-        title = "<br>".join(title_lines)
-
-        fig.update_layout(
-            width=width,
-            height=height,
-            title=title,
-            autosize=True,
-            hovermode="closest",
-            plot_bgcolor="white",
-            yaxis_title="Distance (no. SNPs)",
-            xaxis_title="Haplotypes",
-            showlegend=True,
-        )
+        if min_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples < min_cohort_size:
+                raise ValueError(
+                    f"not enough samples ({n_samples}) for minimum cohort size ({min_cohort_size})"
+                )
+        if max_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples > max_cohort_size:
+                rng = np.random.default_rng(seed=random_seed)
+                loc_downsample = rng.choice(
+                    n_samples, size=max_cohort_size, replace=False
+                )
+                loc_downsample.sort()
+                ds_snps = ds_snps.isel(samples=loc_downsample)
 
-        # Repeat the dataframe so there is one row of metadata for each haplotype
-        df_samples_phased_haps = pd.DataFrame(
-            np.repeat(df_samples_phased.values, 2, axis=0)
-        )
-        df_samples_phased_haps.columns = df_samples_phased.columns
-        # select only columns in hover_data
-        df_samples_phased_haps = df_samples_phased_haps[hover_data]
-        debug("Reorder haplotype metadata to align with haplotype clustering")
-        df_samples_phased_haps = df_samples_phased_haps.loc[
-            fig.layout.xaxis["ticktext"]
-        ]
-        fig.update_xaxes(mirror=False, showgrid=True, showticklabels=False, ticks="")
-        fig.update_yaxes(
-            mirror=False, showgrid=True, showline=True, range=[-2, max_dist + 1]
-        )
+        gt = allel.GenotypeDaskArray(ds_snps["call_genotype"].data)
+        with self._dask_progress(desc="Load genotypes"):
+            gt = gt.compute()
+        pos = ds_snps["variant_position"].values
+
+        if sites in self.site_mask_ids:
+            debug("subsetting to haplotype positions")
+            haplotype_pos = self._haplotype_sites_for_contig(
+                contig=contig,
+                analysis=sites,
+                field="POS",
+                inline_array=True,
+                chunks="native",
+            ).compute()
+            hap_site_mask = np.in1d(pos, haplotype_pos)
```

Comment:
```suggestion
            hap_site_mask = np.in1d(pos, haplotype_pos, assume_unique=True)
```

Apparently can speed up the computation.